linked_list/linked_list.py

```diff
@@ -9,67 +9,167 @@ def __init__(self, value, next_node = None):
     # returns true if found, false otherwise
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def search(self, value):
-        pass
+
+        if self.head == None:
+            return False
+
+        current_node = self.head
+        while current_node.next != None:
+            if current_node.value == value:
+                return True
+            current_node = current_node.next
+        
+        if current_node.value == value:
+            return True
+        return False
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def length(self):
-        pass
+        count = 0
+        current_node = self.head
+
+        while current_node != None:
+            count = count + 1
+            current_node = current_node.next
+        
+        return count
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_at_index(self, index):
-        pass
+        current_node = self.head
+        count = 0
+
+        while current_node != None:
+            if count == index:
+                return current_node.value
+
+            count += 1
+            current_node = current_node.next
+        
+        return None
+
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: o(n)/linear complexity
+    # Space Complexity: o(n)/linear complexity
     def get_last(self):
-        pass
+        
+        #make sure list is not empty
+        if self.head == None:
+            return None
+
+        current_node = self.head
+        while current_node.next != None:
+            current_node = current_node.next
+        
+        return current_node.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head == None: 
+            self.head = new_node
+            return 
+        else:
+            current_node = self.head
+
+            while current_node.next != None:
+                current_node = current_node.next
+                
+            current_node.next = new_node 
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):

```

Comment:
👍 This works, but do you really need to create an array and find the max of the array.  Could you do this with O(1) space complexity?